python/ray/tune/tuner.py

```diff
@@ -0,0 +1,176 @@
+import traceback
+from typing import Any, Callable, Dict, List, Optional, Type, Union
+
+import ray
+
+from ray.tune import TuneError, ExperimentAnalysis
+from ray.tune.api_v2.convertible_to_trainable import ConvertibleToTrainable
+from ray.tune.callback import Callback
+from ray.tune.trainable import Trainable
+from ray.tune.impl.tuner_internal import TunerInternal
+from ray.tune.tuner_config import TuneAlgoConfig
+from ray.util import PublicAPI
+from ray.util.client.common import ClientActorHandle
+from ray.util.ml_utils.node import force_on_current_node
+
+
+########################################################################
+# The motivations behind the new Tuner API as part of Ray MLC effort are that
+# 1. Users can seamlessly transition from their training process to tuning process.
+# 2. Dataset and data preprocessing are now an integral part of training/tuning process.
+#   Bonus point: you can even tune dataset and preprocessing.
+
+# Some requirements on Tuner that may help you understand some implementation decisions:
+# 1. Tuner needs to work in ray client mode.
+# 2. Tuner needs to be restored in case of any failures.
+########################################################################
+
+# The magic key that is used when instantiating Tuner during resume.
+_TUNER_INTERNAL = "tuner_internal"
+_SELF = "self"
+
+
+@PublicAPI(stability="beta")
+class Tuner:
+    """The external facing Tuner class as part of Ray MLC effort.
+
+    Args:
+        trainable (Union[
+                str,
+                Callable,
+                Type[Trainable],
+                Type[ConvertibleToTrainable],
+                ConvertibleToTrainable,
+            ]): The trainable to be tune.
+        param_space (Optional[Dict[str, Any]]): Search space to run tuning on.
+            One thing to note is that both preprocessor and dataset can be tuned here.
+        tune_algo_config (Optional[TuneAlgoConfig]): tune algo specific configs.
+            Refer to ray.tune.tuner_config.TunerAlgoConfig for more info.
+        name (Optional[str]): Name of the run.
+        local_dir (Optional[str]): Local dir to save training results to.
+            Defaults to ``~/ray_results``.
+        callbacks (Optional[List[Callback]]): Callbacks to invoke.
+            Refer to ray.tune.callback.Callback for more info.
+
+    Usage pattern:
+    .. code-block:: python
+
+    # Only do the following, if you want to run in ray client mode.
+    # This means the tune driver code will be running on head node of your cluster.
+    ray.init("ray://my_ray_cluster_address")
```

Comment:
we should start making these doc examples executable. no need to have a pseudo ray cluster address. please add all necessary imports, too. when exposed to docs, this should just run when copied.